### ISSUE TITLE
Built-in opener using environment variables to find appropriate application

### ISSUE THREAD

**Issue by seifrox** [jarun/nnn#798]
First of all thanks for this amazing file manager, I have just stumbled upon recently and i love it!

**Problem/Annoyance**

I am not happy with using `xdg-open` as the default opener, so I tried `nuke` from the plugins which was good but also did not fit my custom needs, of achieving a mixture of CLI and GUI apps for opening, completely. One of the shortcomings of `nuke` for me is that the authors had certain preferences for applications and it is not really extensible. On the other hand I don't want to mix general GUI opening with opening files through `nnn`, so `xdg-open` is also not an option.

**Solution I would like to see**

Following the style of `nnn` configuration, I would like to be able to configure the preferred application via environment variables and use `nnn` as the opener with a configuration similar to the plugins, e.g. to do something like
```
export NNN_OPENER='nnn'

# define openers via extension
# export NNN_OPEN_EXT_<extension>='<cmd>'
# examples:
export NNN_OPEN_EXT_PDF='_|/usr/bin/evince $nnn*'
export NNN_OPEN_EXT_010='_batcat $nnn*'

# define openers via mime type
# export NNN_OPEN_MIME_<mime base type|exact mime type>='<cmd>'
# basically the mime type would have to be sanitized to only contain characters
# allowed in environment variables e.g. 
#         'text/plain' becomes TEXT_PLAIN
#         'video/x-matroska' becomes VIDEO_X_MATROSKA
# the base mime type is basically everything of the mime type up to the first _ character
# examples:
export NNN_OPEN_MIME_TEXT='_batcat $nnn*'
export NNN_OPEN_MIME_TEXT_PLAIN='_less $nnn*'
export NNN_OPEN_MIME_VIDEO='_|mpv $nnn*'
```
The idea would be if `nnn` is used as a opener:

1. I first checks if there is a variable defined for `NNN_OPEN_EXT_<extension>`, given the file extension, if something is defined it will use it to open the file
2. Next it will check if for the exact mime type a variable is defined `NNN_OPEN_MIME_<exact mime type>`, if something is defined it will use it to open the file
3. Next it will check if for the base mime type a variable is defined `NNN_OPEN_MIME_<base mime type>`, if something is defined it will use it to open the file
4. If no definition is `nnn` should write out some message that no application is defined and also give a hint which environment variables could be set to solve the issue

Example: `nnn` is used to open `sample.mkv`

1. existence of `NNN_OPEN_EXT_MKV` is checked
2. existence of `NNN_OPEN_MIME_VIDEO_X_MATROSKA` is checked
3. existence of `NNN_OPEN_MIME_VIDEO` is checked
4. if nothing was defined, print error with hint to either define one of the above environment variables

**Describe alternatives you've considered**

First I fiddled around by creating symbolic links for applications names `nuke` is searching for and I have not installed, to lead `nuke` to my intended application, later when this was not good enough I started fiddling around in the nuke script to specify the application for a file extension unknown to `nuke`.
It just did not feel right, and it was not an elegant solution, nor easy to maintain.

Next, I basically implemented the solution above within a `fish` shell script, and I struggled quite a lot to launch GUI apps without them being closed when the terminal closes. For this I started writing a little `C` program o be called from the script to detach it, and ended up copying some code `nnn` uses to run plugins, since I did not manage to do it on my own.

Basically I have now a solution that works for me, but I don't like the fact that basically `nnn` calls a script which calls a `C` program (which basically copied some code of `nnn`) to finally then launch the intended application. I also don't like the fact that, if I want to share this with my friend I would have to rewrite my script as `bash` script and also share with him my `C` application

**Additional context**
The fish script I wrote
```
#!/usr/bin/env fish
function +open --description 'opens the file at the given path' --argument-names 'file_path'

    function .open --argument-names 'app' 'file_path'
        if test -n "$app"
            set -l gui_app (string match -r '_\|(.+)' "$app")[2]
            if test -n "$gui_app"
                "/usr/bin/guiOpen $gui_app '$file_path'"
            else
                "$app $file_path"
            end
            return 0
        end
        return 1
    end

    if not test -e $file_path
        return 1
    end

    set -l file_name (basename $file_path)
    #prepare extension in upper case to for substitution in the environment variable
    set -l extension (string upper (string split -r -m1 . $file_name)[2])

    set -l key "NNN_OPEN_EXTENSION_$extension"

    if not .open "$$key" "$file_path"
        #prepare mime_type in upper case and replace special characters with _ to substitute the environment variable
        set -l mime_type (string upper (file --dereference --brief --mime-type -- "$file_path"))
        set -l mime_type (string replace -a -r '[^A-Za-z0-9]' '_' "$mime_type")

        set -l key "NNN_OPEN_MIME_$mime_type"

        if not .open "$$key" "$file_path"
            #prepare mime base type
            set -l mime_base_type (string split -m1 '_' $mime_type)[1]
 
            set -l key "NNN_OPEN_MIME_$mime_base_type"

            if not .open "$$key" "$file_path"
                if test -z "$extension"
                    printf "define an application via one of the following environment variables:\nNNN_OPEN_MIME_%s\nNNN_OPEN_MIME_%s" "$mime_type" "$mime_base_type"
                else
                    printf "define an application via one of the following environment variables:\nNNN_OPEN_EXTENSION_%s\nNNN_OPEN_MIME_%s\nNNN_OPEN_MIME_%s" "$extension" "$mime_type" "$mime_base_type"
                end
                return 1
            end
        end
    end
    return 0
end

+open $argv

```

**Consider contributing**

I am not completely sure if you would like to have such a feature within `nnn`, or if it rather should be a plugin, however since I need a `C` program anyway that shares code with `nnn`, I though it would be a good fit.

If you think it would be a good addition, I could try to implement it within `nnn` and submit a pull request. Just for those who would like to see this feature soon and are waiting on me to implement a first version, at the moment I don't have so much time, and the last time I coded in C  was over a decade ago, so I guess getting familiar with the code and figure out how to fit it in will take some time.

**Comment by jarun** [comment]
> my custom needs, of achieving a mixture of CLI and GUI apps for opening, completely

You can use both GUI and CLI apps together in `nuke`. There's a variable to do that. Please read the in-file documentation.

> I would like to be able to configure the preferred application via environment variables

That's what `nuke` does cleanly in a script. We don't want to support/read tons of environment variables.

I think your proposal is adding a new layer of environment variables. Given there is no limit to mimes, I can see how this can easily shoot up the number of environment variables in the system.

> I also don't like the fact that, if I want to share this with my friend I would have to rewrite my script as bash script and also share with him my C application

Systems are different with different applications installed. Whatever you write cannot be portable and would have to be adapted to the new system.

I'm sorry but I don't think this is the right thing to do. Feel free to fork the project and add your idea. `nnn` is quite stable now and I don't see too many merge issues if you can contain your changes to a single function of something (which I think is possible).

**Comment by seifrox** [comment]
> > my custom needs, of achieving a mixture of CLI and GUI apps for opening, completely
> 
> You can use both GUI and CLI apps together in `nuke`. There's a variable to do that. Please read the in-file documentation.

I did, and I used `export GUI=1` just wasn't happy with the default GUI applications chosen and as I wrote further down worked around it, even enhanced the `nuke` script. It just did not feel right to me, also when I discovered [bat](https://github.com/sharkdp/bat) going back to the script and figuring again out where to update it, was not satisfying for me. This is when I came up with the idea to do it the way I described.

By the way, I just remembered something I also did not like about `nuke` (also retested it just now). It has the same issue as my script did have without the additional C application. All the GUI applications started with it, have the shell as a parent process. Since I don't want to remember which terminal to keep around so that e.g. my PDF viewers keep running, I am in trouble since I am now used to close all windows with $mod+Shift+q in the i3wm, which basically sends a kill. In comparison to quit via Ctrl+D, so the shell sticks around until all child processes are finished. But I guess, this might be a very specific problem not relevant to others.

> > I would like to be able to configure the preferred application via environment variables
> 
> That's what `nuke` does cleanly in a script. We don't want to support/read tons of environment variables.
> 
> I think your proposal is adding a new layer of environment variables. Given there is no limit to mimes, I can see how this can easily shoot up the number of environment variables in the system.

I can see how this could happen, and also understand your concern about it. At the moment my current needs are covered with 4 (extension: PDF, MIME: text, video, image), I guess I will end up with more, but not tons of them. I still like the ease of configuration and the flexibility of the approach such as using different applications in different environments. Additionally since it is just another opener option, not mandatory, people cold decide for themselves if they want to deal with that many environment variables.
 
> > I also don't like the fact that, if I want to share this with my friend I would have to rewrite my script as bash script and also share with him my C application
> 
> Systems are different with different applications installed. Whatever you write cannot be portable and would have to be adapted to the new system.

Sure, but if it came out of the box with `nnn` including a portable implementation, only the environment variables would have to be adapted to his needs

> I'm sorry but I don't think this is the right thing to do. Feel free to fork the project and add your idea. `nnn` is quite stable now and I don't see too many merge issues if you can contain your changes to a single function of something (which I think is possible).

Fair enough, no worries. It just was an idea and I wanted to share it and thanks for thinking it through.

Forking, yes is a possibility, of course. Currently, since my time is limited, I'd rather not spend it on a fork only some people will use, especially considering I already have a working solution to achieve the flexibility I desire (even if its not the "optimal" one from my point of view). If you would have been interested in this feature, the motivation would have been higher to do an implementation in `C` and try to figure out how to integrate it into the `nnn` code.

BR,
Alex
